content/post/performance-metrics-histogram.md

```diff
@@ -0,0 +1,245 @@
+
+## Comparing the Two Approaches
+
+The histogram approach is significantly more efficient in both space and time, and it achieves this by use of compression (a histogram is a form of compression since it loses the original data of individual events).
+The efficiency difference becomes important as the total number of values grows; this is true for both network traffic and time as well as storage requirements and query times in the monitoring system.
+
+That is, if we process 100k events per second, and show performance for 5-minute windows (e.g. 300-second windows), this would require storing data for at least 300*100,000=30,000,000 values, sorting it whenever we add data for a new second (and figuring out how to remove data from the outgoing second), and then picking two values by index, the 99th, and 99.9th percentiles. 
+
+For the histogram approach, we store histograms (holding a bin value and a population count value for each bin—two integers per bin), which are small, and we can add them as well as subtract them. This means we can maintain a window histogram, and add the incoming second's histogram as well as subtract the outgoing second's histogram very efficiently without recomputing the sum of the entire set from all the histograms in the window.
+Once updated, we can pick the bins whose maximum values are the _target 99th_ and _target 99.9th_ percentile latencies respectively, and check whether their population sizes or fractions are _greater than_ 99% and 99.9% respectively. This is a much more efficient approach than the sorted values list. However, this approach requires that we define our bins ahead of time to provide us with values of interest. For example, if we care about the percentage of the population below 1ms, we need a bin whose maximum value is 1ms. We cannot infer what percentage is below 1ms if our two nearest bins are 0.5ms and 10ms, for example (other than the percentage must be between the percentage values of each bin).
+
+
+Note that while in the sorted values list approach, we can obtain the latency of _any percentile_, in the histogram approach we instead obtain _the percentile at a predefined latency_. This means that we have the same data, and the same chart (given sufficient resolution), but _the index_ is moved from the percentile axis (in the sorted values list case) to the latency axis (in the histogram case).
+
+## High Throughput and Metrics Saturation
+
+An additional detractor of the sorted values list approach is scalability. The bandwidth and storage cost for the sorted values list approach grows in tandem with the load. So for each additional message processed per second, an additional metric message must be created, sent, stored, and processed. If we achieve our performance goals of 1mm msgs/sec with 99%-ile latencies below 1ms and 99.9%-ile latencies below 1s, we would have a very difficult time maintaining the metrics monitoring infrastructure!
```

Comment:
It seems the answer to my above comment about `network traffic` is in this paragraph.

Same comment as before about the `1s` vs `10ms` threshold for `99.9%`.